state/protocol/protocol_state/mutator.go

```diff
@@ -5,57 +5,247 @@ import (
+//     there are no gaps in the seals.
+//   - The seals guarantee correctness of the sealed execution result, including the contained
+//     service events. This is actively checked by the verification node, whose aggregated
+//     approvals in the form of a seal attest to the correctness of the sealed execution result,
+//     including the contained.
+//
+// Consensus nodes actively verify protocol compliance for any block proposal they receive,
+// including integrity of each seal individually as well as the seals continuously following the
+// fork. Light clients only process certified blocks, which guarantees that consensus nodes already
+// ran those checks and found the proposal to be valid.
+//
+// Details on SERVICE EVENTS:
+// Consider a chain where a service event is emitted during execution of block A.
+// Block B contains an execution receipt for A. Block C contains a seal for block
+// A's execution result.
+//
+//	A <- .. <- B(RA) <- .. <- C(SA)
+//
+// Service Events are included within execution results, which are stored
+// opaquely as part of the block payload in block B. We only validate, process and persist
+// the typed service event to storage once we process C, the block containing the
+// seal for block A. This is because we rely on the sealing subsystem to validate
+// correctness of the service event before processing it.
+// Consequently, any change to the protocol state introduced by a service event
+// emitted during execution of block A would only become visible when querying
+// C or its descendants.
+//
+// Error returns:
+//   - Per convention, the input seals from the block payload have already confirmed to be protocol compliant.
+//     Hence, the service events in the sealed execution results represent the honest execution path.
+//     Therefore, the sealed service events should encode a valid evolution of the protocol state -- provided
+//     the system smart contracts are correct.
+//   - As we can rule out byzantine attacks as the source of failures, the only remaining sources of problems
+//     can be (a) bugs in the system smart contracts or (b) bugs in the node implementation.
+//     A service event not representing a valid state transition despite all consistency checks passing
+//     is interpreted as case (a) and handled internally within the StateMutator. In short, we go into Epoch
+//     Fallback Mode by copying the parent state (a valid state snapshot) and setting the
+//     `InvalidStateTransitionAttempted` flag. All subsequent Epoch-lifecycle events are ignored.
+//   - A consistency or sanity check failing within the StateMutator is likely the symptom of an internal bug
+//     in the node software or state corruption, i.e. case (b). This is the only scenario where the error return
+//     of this function is not nil. If such an exception is returned, continuing is not an option.
+func (m *stateMutator) ApplyServiceEvents(seals []*flow.Seal) error {
```

Comment:
I would suggest to merge this with `handleServiceEvents`. I don't really see any benefit of separating the only functional line https://github.com/onflow/flow-go/blob/f1965e1861dcc2f2e8b23bb60ab36abb758a2635/state/protocol/protocol_state/mutator.go#L73

`handleServiceEvents` already mutates the in-memory state of `stateMutator` (specifically the embedded `ProtocolStateMachine`). I have already used the nicely detailed documentation of `handleServiceEvents` (commit https://github.com/onflow/flow-go/pull/4868/commits/bdde39102ff25421d0a61efa79d429d482396e89) as a basis for documenting [`ApplyServiceEvents`](https://github.com/onflow/flow-go/blob/bdde39102ff25421d0a61efa79d429d482396e89/state/protocol/protocol_state.go#L98-L152) here.

Reply:
I've slightly changed the logic, making function atomic, it's all events are valid or none will be saved in DB(mirrors what state machine does) https://github.com/onflow/flow-go/pull/4868/commits/69842cd282c24c74972d412cb40f2fbff2ff3f62